In [1]:
# Initialize Otter
import otter
grader = otter.Notebook("p6.ipynb")

In [2]:
import p6_test

# Project 6: Airbnb

## Learning Objectives:

In this project, you will demonstrate how to:

* access and utilize data in CSV files,
* process real world datasets,
* use string methods and sorting function / method to order data.

Please go through [lab-p6](https://git.doit.wisc.edu/cdis/cs/courses/cs220/cs220-f22-projects/-/tree/main/lab-p6) before working on this project. The lab introduces some useful techniques related to this project.

## Testing your code:

Along with this notebook, you must have downloaded the file `p6_test.py`. If you are curious about how we test your code, you can explore this file, and specifically the value of the variable `expected_json`, to understand the expected answers to the questions.

## Project Description:

Data Science can help us understand user behavior on online platform services. This project is about the rooms listed on Airbnb. Since 2008, guests and hosts have used Airbnb to expand on traveling possibilities and present a more unique, personalized way of experiencing the world. `airbnb.csv` has data of nearly 50,000 listings on Airbnb from New York City, NY from the year 2019. This file includes a lot of information about the hosts, geographical availability of the listings, and other necessary metrics to make predictions and draw conclusions. You will be using various string manipulation methods that come with Python as well as creating some of your own functions to solve the problems posed. Happy coding!

## Dataset:

A small portion of the dataset `airbnb.csv` you will be working with for this project is reproduced here:

room_id|name|host_id|host_name|neighborhood_group|neighborhood|latitude|longitude|room_type|price|minimum_nights|number_of_reviews|last_review|reviews_per_month|calculated_host_listings_count|availability_365
------|------|------|------|------|------|------|------|------|------|------|------|------|------|------|------|
2539|Clean & quiet apt home by the park|2787|John|Brooklyn|Kensington|40.64749000000001|-73.97237|Private room|149|1|9|2018-10-19|0.21|6|365
2595|Skylit Midtown Castle|2845|Jennifer|Manhattan|Midtown|40.75362|-73.98376999999998|Entire home/apt|225|1|45|2019-05-21|0.38|2|355
3647|THE VILLAGE OF HARLEM....NEW YORK !|4632|Elisabeth|Manhattan|Harlem|40.80902|-73.9419|Private room|150|3|0|||1|365
3831|Cozy Entire Floor of Brownstone|4869|LisaRoxanne|Brooklyn|Clinton Hill|40.68514|-73.95976|Entire home/apt|89|1|270|2019-07-05|4.64|1|194
5022|Entire Apt: Spacious Studio/Loft by central park|7192|Laura|Manhattan|East Harlem|40.79851|-73.94399|Entire home/apt|80|10|9|2018-11-19|0.1|1|0
5099|Large Cozy 1 BR Apartment In Midtown East|7322|Chris|Manhattan|Murray Hill|40.74767|-73.975|Entire home/apt|200|3|74|2019-06-22|0.59|1|129

You can find more details on the dataset in [lab-p6](https://git.doit.wisc.edu/cdis/cs/courses/cs220/cs220-f22-projects/-/tree/main/lab-p6).

## Questions and Functions:

Let us start by importing all the modules we will need for this project.

In [3]:
# it is considered a good coding practice to place all import statements at the top of the notebook
# please place all your import statements in this cell if you need to import any more modules for this project
import csv
import math

#### Now, copy and paste the `process_csv` and `cell` functions from your [lab-p6](https://git.doit.wisc.edu/cdis/cs/courses/cs220/cs220-f22-projects/-/tree/main/lab-p6) notebook to the cell below.

You are expected to call the `process_csv` function correctly, and read the data on `airbnb.csv`. After reading the file, define the `csv_header`, and `csv_rows` variables as in [lab-p6](https://git.doit.wisc.edu/cdis/cs/courses/cs220/cs220-f22-projects/-/tree/main/lab-p6), and define the `cell` function.

**Important:** You **must** only use the `cell` function to extract data from the dataset. If you extract any data without explicitly using this function, you will **lose points** during manual review. It is recommended but **optional** that you use the `cell_v2` function defined in lab-p6. However, you **must** rename the function to `cell` in your p6 notebook. 

In [4]:
# modified from https://automatetheboringstuff.com/chapter14/
def process_csv(filename):
    example_file = open(filename, encoding="utf-8")
    example_reader = csv.reader(example_file)
    example_data = list(example_reader)
    example_file.close()
    return example_data

In [5]:
def cell(row_idx, col_name):
    """
    cell(row_idx, col_name) returns the data value (cell) 
    corresponding to the row index `row_idx`
    and the column name `col_name` of a CSV file
    """
    # replace the ... with your code
    col_idx = csv_header.index(col_name)
    val = csv_rows[row_idx][col_idx]
    if val == "": # when we come across missing data, we return None instead
        return None
    return val

In [6]:
csv_rows = process_csv("airbnb.csv")
csv_header = csv_rows[0]
csv_data = csv_rows[1:]

In [7]:
csv_rows


[['room_id',
  'name',
  'host_id',
  'host_name',
  'neighborhood_group',
  'neighborhood',
  'latitude',
  'longitude',
  'room_type',
  'price',
  'minimum_nights',
  'number_of_reviews',
  'last_review',
  'reviews_per_month',
  'calculated_host_listings_count',
  'availability_365'],
 ['2539',
  'Clean & quiet apt home by the park',
  '2787',
  'John',
  'Brooklyn',
  'Kensington',
  '40.64749000000001',
  '-73.97237',
  'Private room',
  '149',
  '1',
  '9',
  '2018-10-19',
  '0.21',
  '6',
  '365'],
 ['2595',
  'Skylit Midtown Castle',
  '2845',
  'Jennifer',
  'Manhattan',
  'Midtown',
  '40.75362',
  '-73.98376999999998',
  'Entire home/apt',
  '225',
  '1',
  '45',
  '2019-05-21',
  '0.38',
  '2',
  '355'],
 ['3647',
  'THE VILLAGE OF HARLEM....NEW YORK !',
  '4632',
  'Elisabeth',
  'Manhattan',
  'Harlem',
  '40.80902',
  '-73.9419',
  'Private room',
  '150',
  '3',
  '0',
  '',
  '',
  '1',
  '365'],
 ['3831',
  'Cozy Entire Floor of Brownstone',
  '4869',
  'LisaRoxanne'

**Question 1:** What **unique** `neighborhood_groups` are included in the dataset?

Your output **must** be a *list* which stores all the **unique** neighborhood groups (i.e., without any duplicates). The order **does not** matter.

In [10]:
# compute and store the answer in the variable 'neighborhood_groups', then display it
neighborhood_groups = []
for row_idx in range(len(csv_rows)-1):
    neighborhood_group = cell(row_idx, "neighborhood_group")
#     if neighborhood_group != None:
    neighborhood_groups.append(neighborhood_group)
    neighborhood_groupsSet = set(neighborhood_groups)
    neighborhood_groups = list(neighborhood_groupsSet)
    
neighborhood_groups.remove(neighborhood_groups[2])

neighborhood_groups 

['Brooklyn', 'Queens', 'Staten Island', 'Manhattan', 'Bronx']

In [11]:
grader.check("q1")

q1 results: All test cases passed!

**Question 2:** What is the **average** `price` of rooms in the dataset?

In [36]:
#compute and store the answer in the variable 'avg_price', then display it
total = 0
avg_price = 0
for row_idx in range(len(csv_data)):
    avg_price+=int(cell(row_idx+1, "price"))

avg_price = avg_price/len(csv_rows)
avg_price
    


152.71756380890054

In [37]:
grader.check("q2")

q2 results: All test cases passed!

**Question 3:** How many rooms are in the `neighborhood` of *SoHo*?

In [38]:
# compute and store the answer in the variable 'count_soho', then display it
# replace the ... with your code
# replace the ... with your code
count_soho = 0 # initialize with the correct value
for row_idx in range(len(csv_data)): # loop through all the indices 
    if cell(row_idx, "neighborhood") == "SoHo": # use `cell` to determine if the listing at `idx` is from the correct neighborhood_group
        count_soho += 1 # update the variable appropriately

count_soho

358

In [39]:
grader.check("q3")

q3 results: All test cases passed!

### Function 1: `find_room_names(phrase)`

We require you to complete the below function to answer the next several questions (this is a **requirement**, and you will **lose points** if you do not implement this function).

In [40]:
def find_room_names(phrase):
    """
    find_room_names(phrase) returns a list of all the room names that contain the 
    substring (case insensitive match) `phrase`.
    """
    rooms = []
    
    for row_idx in range(len(csv_data)):
        room = cell(row_idx+1, "name")
        if room is not None:
            if phrase.lower() in room.lower():
                rooms.append(room) 
            
    return rooms
       
    # TODO: create a list
    # TODO: ignore rooms that do not have data entry for name, as indicated by a value of None
    # TODO: check if the room name string contains phrase (case insensitive match)
    # TODO: if so, add these room names to the list (the room names should be as in the dataset)
    # TODO: return your list of room names

**Question 4:** Find all room `names` that contain the string `"CBG"`.
    
Your output **must** be a *list*. The order **does not** matter. You **must** call the `find_room_names` function to answer this question.

In [41]:
# compute and store the answer in the variable 'rooms_contain_cbg', then display it
room_name = find_room_names("CBG")
rooms_contain_cbg = room_name
rooms_contain_cbg

['CBG CtyBGd HelpsHaiti rm#1:1-4',
 'CBG Helps Haiti Room#2.5',
 'CBG Helps Haiti Rm #2',
 'CBG# 4Tiny room w/ huge window/AC',
 'CBG Helps Haiti Rm #3',
 'CBG HelpsHaiti #5 Suite']

In [42]:
grader.check("q4")

q4 results: All test cases passed!

**Question 5:** Find all room `names` that contain **either** `"cinema"` **or** `"film"`.

Your output **must** be a *list*. The order **does not** matter, but if a room's `name` contains **both** `"cinema"` and `"film"`, then the room must be included **only once** in your list. You **must** call the `find_room_names` function to answer this question.

In [43]:
# compute and store the answer in the variable 'rooms_contain_cinema_film', then display it
cinema_room = find_room_names("Cinema")
film_room = find_room_names("Film")

rooms_contain_cinema_film = film_room + cinema_room

rooms_contain_cinema_film = list(set(rooms_contain_cinema_film))
rooms_contain_cinema_film

["Downtown Filmmaker's Loft by WTC",
 'The Otheroom Bar/Event/Filming Space -read details',
 'Cinema + gym included with room',
 'Shoot. Film. Sleep. Unique Loft Space in Brooklyn.',
 'HUGE LUX 2FLOOR 2 BDRMSOHO LOFTw/HOME CINEMA',
 'Cinema Studio on Duplex Apt.',
 'WoodyAllen FilmSet-Like Digs (Room)',
 'Clean music/film themed bedroom',
 'Premium Chelsea 1BR w/ Gym, W/D, Doorman, Sundeck, Cinema, by Blueground',
 'Modern Townhouse for Photo, Film &  Daytime Events',
 'Brooklyn townhouse for filming',
 'Cool apartment in Brooklyn with free cinema & gym',
 'Sunny private room featured in film',
 'Film / photography location in unique apartment',
 'TV-PHOTO-FILM-CINEMA-ART GALLERY-MUSIC STUDIO-LOFT',
 'WoodyAllen FilmSet-Like Digs (Apt)',
 'Victorian Film location',
 'Film Location',
 'Music Recording Mixing Film Photography Art',
 'Stunning Chelsea 1BR w/ Gym, W/D, Doorman, Sundeck, Cinema, by Blueground']

In [44]:
grader.check("q5")

q5 results: All test cases passed!

**Question 6:** Find the **longest** room `name` that contains the word `"fun"`.

There is a **unique** such room with the longest `name`, so you **do not** have to worry about breaking ties. You **must** call the `find_room_names` function to answer this question.

In [45]:
# compute and store the answer in the variable 'funnest_room', then display it
fun_room = find_room_names("Fun")
curr_high = 0
for room in fun_room:
    if len(room) > curr_high:
        curr_high = len(room)
        funnest_room = room

funnest_room

'Homey 1BR in Fun, Central West Village w/ Doorman by Blueground'

In [46]:
grader.check("q6")

q6 results: All test cases passed!

**Question 7:** Find all the `names` of the rooms which have `price` *0* and have **more than** *30* `reviews`.

Your output **must** be a *list*. The names **must** be sorted in **ascending (alphabetical) order**.

In [47]:
# compute and store the answer in the variable 'no_cost_rooms', then display it
no_cost_rooms = []
num_reviews = 30
low = 0 

for idx in range(len(csv_data)):
    price = cell(idx+1, "number_of_reviews")
#     if cell(idx, "number_of_reviews") != None:
    if int(price) > num_reviews and int(cell(idx+1, "price")) == 0:
        no_cost_rooms.append(cell(idx+1, "name"))
            
no_cost_rooms = sorted(no_cost_rooms)
no_cost_rooms

['Contemporary bedroom in brownstone with nice view',
 'Cozy yet spacious private brownstone bedroom',
 'Spacious comfortable master bedroom with nice view',
 '★Hostel Style Room | Ideal Traveling Buddies★']

In [48]:
grader.check("q7")

q7 results: All test cases passed!

**Question 8:**  What `neighborhoods` are the rooms that have `prices` greater than *9999* located in?

Your output **must** be a *list* of **unique** neighborhoods (i.e., without any duplicates). The names **must** be sorted in **descending (reverse-alphabetical) order**.

In [49]:
pricey_neighborhoods = []
pricey_price = 9999

for idx in range(len(csv_data)):
    prices = cell(idx+1, "price")
#     if cell(idx, "price") != None:
 
    if int(prices) > pricey_price:
        pricey_neighborhoods.append(cell(idx+1, 'neighborhood'))
        

pricey_neighborhoods.sort(reverse=True)
pricey_neighborhoods

['Upper West Side', 'Greenpoint', 'Astoria']

In [50]:
grader.check("q8")

q8 results: All test cases passed!

**Question 9:** How many rooms received their `last_review` in or before *2015*?

You should **ignore** rooms for which the `last_review` data is missing.

**Hint:** You can find the date of the last review in the `last_review` column.

In [51]:
# compute and store the answer in the variable 'last_review_before_2015', then display it
last_review_before_2015 = 0

for idx in range(len(csv_rows)):
    if cell(idx, "last_review") is not None:
        if cell(idx, "last_review")[:-6] <= '2015':
            last_review_before_2015 += 1

last_review_before_2015
    
    

1672

In [52]:
grader.check("q9")

q9 results: All test cases passed!

### Function 2: `avg_price_per_room_type(room_type, neighborhood)`

We require you to complete the below function to answer the next several questions (this is a **requirement**, and you will **lose points** if you do not implement this function).

In [53]:
def avg_price_per_room_type(room_type, neighborhood):
    '''
    avg_price_per_room_type(room_type, neighborhood) returns the average price of 
    rooms of the type `room_type` in the given `neighborhood`; if there are no
    rooms of the type `room_type` in the given `neighborhood`, it returns `None`
    '''
    price_of_rooms = 0
    num_of_room_types = 0
    for row_idx in range(len(csv_data)):
        if cell(row_idx, "neighborhood") is not None and cell(row_idx, "neighborhood") == neighborhood:
            
            if room_type == cell(row_idx, "room_type"):
                num_of_room_types += 1   
                
                price_of_rooms += float(cell(row_idx, "price"))           
    if num_of_room_types == 0:
        return None
        
    return price_of_rooms / num_of_room_types

**Question 10:** What is the  **average** cost of a *Private room* in the`neighborhood` *Little Neck*?

You **must** call the `avg_price_per_room_type` function to answer this question.

**Hint:** To help you debug your code in case you run into any bugs, we have reproduced in the cell below, **all** the rows in the dataset from the `neighborhood` *Little Neck*. If you run into bugs with `avg_price_per_room_type`, it is recommended that you go through your code and verify that it does what it is supposed to, for this tiny dataset.

room_id|name|host_id|host_name|neighborhood_group|neighborhood|latitude|longitude|room_type|price|minimum_nights|number_of_reviews|last_review|reviews_per_month|calculated_host_listings_count|availability_365
------|------|------|------|------|------|------|------|------|------|------|------|------|------|------|------|
20227428|Douglaston Apartment Room A|18996093|Leonard|Queens|Little Neck|40.75794000000001|-73.72955999999998|Private room|45|1|12|2019-06-22|0.55|5|133
21025083|Douglaston (apt 2) Room one\n(Largest room)|18996093|Leonard|Queens|Little Neck|40.75777|-73.72949|Private room|50|1|6|2018-12-16|0.31|5|94
30325639|Cozy shared studio in a safe neighborhood|21495656|Ramy|Queens|Little Neck|40.76212|-73.71928|Shared room|32|3|1|2018-12-04|0.14|1|88
31553066|Near major transportation|41090359|Abi|Queens|Little Neck|40.77122|-73.738|Private room|100|1|0|||1|88
35515780|30-min to Manhattan Quiet Big House in Great Neck|31859704|Vincent|Queens|Little Neck|40.77444000000001|-73.73373000000002|Entire home/apt|149|3|0|||1|3

In [54]:
# compute and store the answer in the variable 'pvt_room_little_neck', then display it
pvt_room_little_neck = avg_price_per_room_type("Private room", "Little Neck")
pvt_room_little_neck

65.0

In [55]:
grader.check("q10")

q10 results: All test cases passed!

**Question 11:** How much more expensive is a *Entire home/apt* than a *Private room* in the `neighborhood` *Astoria*?

You **must** call the `avg_price_per_room_type` function to answer this question.

In [56]:
# compute and store the answer in the variable 'home_pvt_room_astoria_diff', then display it
avg_home_and_apt = avg_price_per_room_type("Entire home/apt", "Astoria")
private_room_astoria = avg_price_per_room_type("Private room", "Astoria")
home_pvt_room_astoria_diff = float(avg_home_and_apt - private_room_astoria)
home_pvt_room_astoria_diff

46.02133741379495

In [57]:
grader.check("q11")

q11 results: All test cases passed!

### Function 3: `find_prices_within(lat_min, lat_max, long_min, long_max)` 

We require you to complete the below function to answer the next several questions (this is a **requirement**, and you will **lose points** if you do not implement this function).

In [60]:
def find_prices_within(lat_min, lat_max, long_min, long_max):
    """
    find_prices_within(lat_min, lat_max, long_min, long_max) returns an unordered 
    list of prices of all the rooms within the geographical location between and including
    the latitudes lat_min and lat_max and longitudes long_min and long_max.
    """

    
    for idx in range(len(csv_data)):
        latitude = float(cell(idx, 'latitude'))
        longitude = float(cell(idx, 'longitude'))
        if long_min <= longitude <= long_max and lat_min <= latitude <= lat_max:
            
            print(latitude, longitude)
        
        
        
find_prices_within(40.729, 40.73, -74.01, -74.00)

ValueError: could not convert string to float: 'latitude'

In [59]:
def avg_price_per_room_type(room_type, neighborhood):
    '''
    avg_price_per_room_type(room_type, neighborhood) returns the average price of 
    rooms of the type `room_type` in the given `neighborhood`; if there are no
    rooms of the type `room_type` in the given `neighborhood`, it returns `None`
    '''
    price_of_rooms = 0
    num_of_room_types = 0
    for row_idx in range(len(csv_data)):
        if cell(row_idx, "neighborhood") is not None and cell(row_idx, "neighborhood") == neighborhood:
            
            if room_type == cell(row_idx, "room_type"):
                num_of_room_types += 1   
                
                price_of_rooms += float(cell(row_idx, "price"))           
    if num_of_room_types == 0:
        return None
        
    return price_of_rooms / num_of_room_types

**Question 12:** What is the **lowest** `price` room near *NYU* (`40.729 <= latitude <= 40.73, -74.01 <= longitude <= -74.00`)?

You **must** call the `find_prices_within` function to answer this question.

In [ ]:
# compute and store the answer in the variable 'min_price_nyu', then display it
min_price_nyu = find_prices_within(40.729, 4073, -74.01, -74)
min_price_nyu

In [ ]:
grader.check("q12")

### Function 4: `median(items)` 

We require you to complete the below function to answer the next several questions (this is a **requirement**, and you will **lose points** if you do not implement this function).

In [ ]:
def median(items):
    """
    median(items) returns the median of the list `items`
    """
    pass # replace with your code
    # you may copy/paste this function from your lab-p6 notebook

**Question 13:** What is the **median** `price` of the rooms near *Columbia University* (`40.79 <= latitude <= 40.80, -73.96 <= longitude <= -73.95`)?

You **must** call the `find_prices_within` function to answer this question.

In [ ]:
# compute and store the answer in the variable 'median_price_columbia', then display it


In [ ]:
grader.check("q13")

**Question 14:** What **percentage** of rooms near *Rockefeller Center* (`40.749 <= latitude <= 40.75, -73.98 <= longitude <= -73.97`) have a `price` **more than** *100*?

Your answer **must** be a *float* value between *0* and *100*. You **must** call the `find_prices_within` function to answer this question.

In [ ]:
# compute and store the answer in the variable 'pct_price_over_hundred', then display it


In [ ]:
grader.check("q14")

### Function 5: `review_avail_ratio(neighborhood)` 

We require you to complete the below function to answer the next several questions (this is a **requirement**, and you will **lose points** if you do not implement this function).

You should ignore rooms that have `availability_365` data of 0. You should also ignore rooms for which the ratio cannot be computed due to missing data.

1.  The denominator is the availability of a room (`availability_365` column). The numerator is the number of reviews of a room (`number_of_reviews column`). 
2.  Be careful! You need to compute the ratio for each room in the given neighborhood, then take the average of those ratios. Simply dividing the sum of reviews by the sum of availability will calculate the wrong answer.


In [ ]:
def review_avail_ratio(neighborhood):
    """
    review_avail_ratio(neighborhood) returns the average of the ratios of 
    number of reviews to availability of all rooms in the `neighborhood`
    """
    pass # replace with your code
    # TODO: you should **ignore** rooms that have `availability_365` data of 0
    # TODO: you should **ignore** rooms for which the ratio cannot be computed due to missing data
    # Hint: the numerator is the number of reviews of a room (`number_of_reviews column`)
    # Hint: the denominator is the availability of a room (`availability_365` column)
    # Hint: note that you need to compute the average of the ratios, **not** the ratio of the averages.
    #       you must compute the ratio for each room in the `neighborhood`, then take the average of those ratios.
    #       simply dividing the sum of reviews by the sum of availability will calculate the wrong answer.

**Question 15:** What is the **average of the ratios** of the `number_of_reviews` to `availability_365` in the `neighborhood` *Bushwick*?

You **must** call the `review_avail_ratio` function to answer this question.

In [ ]:
# compute and store the answer in the variable 'bushwick_avg_ratio', then display it


In [ ]:
grader.check("q15")

**Question 16:** What is the **average of the ratios** of the `number_of_reviews` to `availability_365` in the `neighborhood` *Manhattan Beach*?

You **must** call the `review_avail_ratio` function to answer this question.

In [ ]:
# compute and store the answer in the variable 'manhattan_beach_avg_ratio', then display it


In [ ]:
grader.check("q16")

**Question 17:** Which `neighborhood` in the `neighborhood_group` *Staten Island* has the **highest average of ratios** of the `number_of_reviews` to `availability_365`?

**Clarification:** Don't worry about it if this cell takes around 10 seconds to run, that is expected. If it takes much longer (i.e., more than 30 seconds), you **must** optimize your code. Attend office hours if you are unable to get your code to run faster.

**Hint:** You do not need to compute the average of ratios for the **same** `neighborhood` more than once. Make a list of the **unique** neighborhoods in *Staten Island* first, and then find the **highest average of ratios** among those `neighborhoods`.

In [ ]:
# compute and store the answer in the variable 'max_nbhd_staten_island', then display it


In [ ]:
grader.check("q17")

### Function 6: `secondary_word_in_found_rooms(find_room_word, secondary_word)`


We require you to complete the below function to answer the next several questions (this is a **requirement**, and you will **lose points** if you do not implement this function).

`secondary_word_in_found_rooms` function definition **must** invoke the function `find_room_names`. **We'll manually deduct points** if you don't use `find_room_names`.

In [ ]:
def secondary_word_in_found_rooms(find_room_word, secondary_word):
    """
    secondary_word_in_found_rooms(find_room_word, secondary_word) returns 
    the percentage of names containing the word `find_room_word` (case insensitive match)
    that also contains the word `secondary_word` (case insensitive match).
    """    
    pass # replace with your code
    # Hint: the denominator is the number of rooms with `find_room_word` in their name.
    # Hint: the numerator is the number of rooms that have **both** `find_room_word` and `secondary_word` in their name.

**Question 18:** What **percentage** of rooms whose `names` contain the word `"downtown"` **also** contain the word `"spacious"`?

Your answer **must** be a *float* value between *0* and *100*. You **must** call the `secondary_word_in_found_rooms` function to answer this question.

In [ ]:
# compute and store the answer in the variable 'downtown_and_spacious', then display it


In [ ]:
grader.check("q18")

**Question 19:** What **percentage** of rooms whose `names` contain the word `"university"` **also** contain the word `"studio"`?

Your answer **must** be a *float* value between *0* and *100*. You **must** call the `secondary_word_in_found_rooms` function to answer this question.

In [ ]:
# compute and store the answer in the variable 'university_and_studio', then display it


In [ ]:
grader.check("q19")

**Question 20:** On a trip to NYC, you need to stay for *3* days in *Queens*, and then *4* days in *Brooklyn*. What is the **minimum** amount of money you need to spend on this trip?

Note that:
1. The price of each room is for each day.
2. The total cost = (lowest price in Queens) * 3 + (lowest price in Brooklyn) * 4.
3. You'll need to **skip** those rooms that don't have enough availability.


**Hint:** You might want to define a helper function to compute the **minimum** daily `price` of a room in a given `neighborhood_group` among rooms whose `availability_365` is equal to or larger than the number of days one will be staying in that neighborhood group.

In [ ]:
# compute and store the answer in the variable 'min_cost_trip', then display it


In [ ]:
grader.check("q20")

## Submission

Make sure you have run all cells in your notebook in order before running the cell below, so that all images/graphs appear in the output. The cell below will generate a zip file for you to submit. **Please save before exporting!**

**SUBMISSION INSTRUCTIONS**: 1. **Save** the notebook file **now (before you run the next cell of code)**. 2. **Upload** the zipfile to Gradescope. 3. Check **Gradescope otter** results as soon as the auto-grader execution gets completed. Don't worry about the score showing up as -/100.0. You only need to check that the test cases passed.

In [ ]:
# Save your notebook first, then run this cell to export your submission.
grader.export(pdf=False, run_tests=True)